allennlp/modules/span_extractors/locally_normalised_span_extractor.py

```diff
@@ -0,0 +1,96 @@
+
+from allennlp.common.params import Params
+from allennlp.modules.span_extractors.span_extractor import SpanExtractor
+from allennlp.modules.time_distributed import TimeDistributed
+from allennlp.nn import util
+
+@SpanExtractor.register("locally_normalised")
+class LocallyNormalisedSpanExtractor(SpanExtractor):
+    """
+    Computes span representations by generating an unnormalized attention score for each
+    word in the document, compute distributions over every span with respect to these
+    scores by normalising the attention scores for words inside the span.
+
+    Given these attention distributions over every span, weight the corresponding vector
+    representations of the words in the span by this distribution, returning a weighted
+    representation of each span.
+
+    Parameters
+    ----------
+    input_dim : ``int``, required.
+        The last dimension of the sequence representations.
+
+    Returns
+    -------
+    attended_text_embeddings : ``torch.FloatTensor``
+        A tensor of shape (batch_size, num_spans, input_dim), which each span representation
+        is formed by locally normalising a global attention over the sequence. The only way
+        in which the attention distribution differs over different spans is in the set of words
+        over which they are normalized.
+    """
+    def __init__(self,
+                 input_dim: int) -> None:
+        super().__init__()
+        self._global_attention = TimeDistributed(torch.nn.Linear(input_dim, 1))
+
+    @overrides
+    def forward(self, # pylint: disable=arguments-differ
+                sequence_tensor: torch.FloatTensor,
+                indicies: torch.LongTensor) -> None:
+        # both of shape (batch_size, num_spans, 1)
+        span_starts, span_ends = indicies.split(1, dim=-1)
+
+        # shape (batch_size, num_spans, 1)
+        span_widths = span_ends - span_starts
+
+        # We need to know the maximum span width so we can
+        # generate indices to extract the spans from the sequence tensor.
+        # These indices will then get masked below, such that if the length
+        # of a given span is smaller than the max, the rest of the values
+        # are masked.
+        max_batch_span_width = int(span_widths.max().data) + 1
+
+        # shape (batch_size, sequence_length, 1)
+        global_attention_logits = self._global_attention(sequence_tensor)
+
+        # Shape: (1, 1, max_batch_span_width)
+        max_span_range_indices = util.get_range_vector(max_batch_span_width,
+                                                       sequence_tensor.is_cuda).view(1, 1, -1)
+        # Shape: (batch_size, num_spans, max_batch_span_width)
+        # This is a broadcasted comparison - for each span we are considering,
+        # we are creating a range vector of size max_span_width, but masking values
+        # which are greater than the actual length of the span.
+        span_mask = (max_span_range_indices <= span_widths).float()
```

Comment:
This `<=` also made me think a whole lot, along with the `>= 0`.  I was wondering why there weren't any off-by-one errors here.  Your excellent test convinced me that the logic was correct, but then I had to figure out why it was correct.  Because the span starts and ends are inclusive, your `span_widths` calculation above is actually off by one, and you compensate for that here with a `<=` instead of a `<`.  The fact that `raw_span_indices` actually gives the indices _backwards_ also confused me a bit.  There was a lot going on in these few lines... =).

I'm not sure what the best thing to do here is, but at the least some more detailed comments would be nice, because this is a potentially pretty confusing block of code.  E.g., if you want to keep the off-by-one issue in `span_widths`, that's fine, as long as there's a note describing this.  Might also be worth stating (even with an example) that `raw_span_indices` will be backwards, so if you're doing a computation that depends on order, you might be surprised.

Reply:
The `raw_span_indices` are actually permutation invariant, because we use the indices generated to select both the embeddings _and_ the attention logits, so because they are the same for both, the order doesn't matter for any given span.

I agree that it's not ideal that this depends on having end-inclusive indices. I'm going to leave a TODO to add a flag to support both I think.